File: src/types/history.ts
import type {
  AuthConfig,
  BodyConfig,
  HttpErrorKind,
  HttpMethod,
  KeyValueRow,
} from "./http";

/**
 * Full standalone snapshot of a request as it was actually sent, plus response
 * metadata. Deliberately NOT a pointer to a saved Request row: ad-hoc/unsaved
 * requests still show up here, and editing a saved request later never
 * rewrites what history shows for a past run.
 */
export interface HistoryEntry {
  readonly id: string;
  readonly method: HttpMethod;
  readonly url: string;
  readonly params: readonly KeyValueRow[];
  readonly headers: readonly KeyValueRow[];
  readonly auth: AuthConfig;
  readonly body: BodyConfig;
  readonly statusCode: number | null;
  readonly durationMs: number | null;
  readonly responseSizeBytes: number | null;
  readonly responseHeaders: readonly (readonly [string, string])[] | null;
  readonly errorKind: HttpErrorKind | null;
  readonly executedAt: string;
}

File: src-tauri/src/models/history.rs
use serde::{Deserialize, Serialize};

use super::request::{AuthPayload, KeyValueRow, StoredBody};

/// Full standalone snapshot of a request as sent, plus response metadata
/// (no response body — see the history design decision). Never a pointer
/// to a saved Request row.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct HistoryEntry {
    pub id: String,
    pub method: String,
    pub url: String,
    pub params: Vec<KeyValueRow>,
    pub headers: Vec<KeyValueRow>,
    pub auth: AuthPayload,
    pub body: StoredBody,
    pub status_code: Option<i64>,
    pub duration_ms: Option<i64>,
    pub response_size_bytes: Option<i64>,
    pub response_headers: Option<Vec<(String, String)>>,
    pub error_kind: Option<String>,
    pub executed_at: String,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct HistoryEntryInput {
    pub method: String,
    pub url: String,
    pub params: Vec<KeyValueRow>,
    pub headers: Vec<KeyValueRow>,
    pub auth: AuthPayload,
    pub body: StoredBody,
    pub status_code: Option<i64>,
    pub duration_ms: Option<i64>,
    pub response_size_bytes: Option<i64>,
    pub response_headers: Option<Vec<(String, String)>>,
    pub error_kind: Option<String>,
}
